File: 146-lru-cache/lru-cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.stack = []

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        else:
            self.stack.remove(key)
            self.stack.append(key)

            return self.cache[key]
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache[key] = value

            self.stack.remove(key)
            self.stack.append(key)
        else:
            if len(self.cache) < self.capacity:
                self.cache[key] = value
                self.stack.append(key)
            else:
                lru_key = self.stack.pop(0)
                del self.cache[lru_key]

                self.cache[key] = value
                self.stack.append(key)
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/lru-cache.java
class LRUCache {
    private int capacity;
    private HashMap<Integer, Integer> cache;
    private ArrayList<Integer> stack;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>();
        this.stack = new ArrayList<>();
    }
    
    public int get(int key) {
        if (!cache.containsKey(key)){
            return -1;
        }
        else {
            stack.remove(Integer.valueOf(key));
            stack.add(key);
            return cache.get(key);
        }
    }
    
    public void put(int key, int value) {
        if (cache.containsKey(key)){
            cache.put(key, value);
            stack.remove(Integer.valueOf(key));
            stack.add(key);
        }
        else{
            if (cache.size() < capacity){
                cache.put(key, value);
                stack.add(key);
            }
            else{
                int lru_key = stack.remove(0);
                cache.remove(lru_key);

                cache.put(key, value);
                stack.add(key);
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */